File: packages/plugin-swc/src/generate_hash.rs
/**
 * KEEP IN SYNC:
 * - `packages/message-utils/src/generate-hash.ts`
 * - `packages/swc-plugin/src/generate_hash.rs`
 */
use data_encoding::BASE64;
use sha2::{Digest, Sha256};

///
/// Generates a unique hash from any string, can be used as a translation key.
///
pub fn generate_hash(input: &str, context: Option<&str>) -> String {
  let mut hasher = Sha256::new();

  hasher.update(input.as_bytes());
    hasher.update([0x1F]);
  hasher.update(context.unwrap_or("").as_bytes());

  let result = hasher.finalize();
  BASE64.encode(result.as_ref())[..6].to_string()
}

File: packages/message-utils/src/generate-hash.ts
/**
 * KEEP IN SYNC:
 * - `packages/message-utils/src/generate-hash.ts`
 * - `packages/swc-plugin/src/generate_hash.rs`
 */

import { sha256 } from 'js-sha256';

/**
 * Generates a unique hash from any string, can be used as a translation key.
 */
export function generateHash(input: string, context?: string): string {
  const hasher = sha256.create();
  if (context) {
    hasher.update(`${input}\u{001F}${context}`);
  } else {
    hasher.update(input);
  }

  const result = hasher.toString();
  const elements = result.match(/.{1,2}/g)?.map((b) => parseInt(b, 16)) || [];
  const bytes = Uint8Array.from(elements);
  return btoa(String.fromCharCode(...bytes)).slice(0, 6);
}
